index out of range in etchosts.Delete

In certain cases there is a docker service restart due to an "index out of range" runtime error when using weave net with mesos/marathon to deploy containers.


Log:

Container 8812916ec6104b2706396c779434d31cb10574160a223364c2c57f7135ace860 failed to exit within 0 seconds of SIGTERM - using the force 
panic: runtime error: index out of range

goroutine 315 [running]:
github.com/docker/libnetwork/etchosts.Delete(0xc8216ebc00, 0x61, 0xc820a721c0, 0x1, 0x1, 0x0, 0x0)
	/usr/src/docker/vendor/src/github.com/docker/libnetwork/etchosts/etchosts.go:154 +0x543
github.com/docker/libnetwork.(*sandbox).deleteHostsEntries(0xc820d1bc00, 0xc820a721c0, 0x1, 0x1)
	/usr/src/docker/vendor/src/github.com/docker/libnetwork/sandbox.go:815 +0x6c
github.com/docker/libnetwork.(*endpoint).sbLeave(0xc820085900, 0x7ff74622eae0, 0xc820d1bc00, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/usr/src/docker/vendor/src/github.com/docker/libnetwork/endpoint.go:599 +0xb35
github.com/docker/libnetwork.(*endpoint).Leave(0xc821681100, 0x7ff74622eae0, 0xc820d1bc00, 0x0, 0x0, 0x0, 0x0, 0x0)
	/usr/src/docker/vendor/src/github.com/docker/libnetwork/endpoint.go:536 +0x1bf
github.com/docker/libnetwork.(*sandbox).delete(0xc820d1bc00, 0xc8215b7b00, 0x0, 0x0)
	/usr/src/docker/vendor/src/github.com/docker/libnetwork/sandbox.go:199 +0xad4
github.com/docker/libnetwork.(*sandbox).Delete(0xc820d1bc00, 0x0, 0x0)
	/usr/src/docker/vendor/src/github.com/docker/libnetwork/sandbox.go:166 +0x32
github.com/docker/docker/daemon.(*Daemon).releaseNetwork(0xc820156780, 0xc820d1b500)
	/usr/src/docker/.gopath/src/github.com/docker/docker/daemon/container_operations_unix.go:1013 +0x4f1
github.com/docker/docker/daemon.(*Daemon).Cleanup(0xc820156780, 0xc820d1b500)
	/usr/src/docker/.gopath/src/github.com/docker/docker/daemon/start.go:159 +0x3d
github.com/docker/docker/container.(*containerMonitor).Close(0xc8215d0680, 0x0, 0x0)
	/usr/src/docker/.gopath/src/github.com/docker/docker/container/monitor.go:127 +0x68
github.com/docker/docker/container.(*containerMonitor).start.func1(0xc8216cbd5f, 0xc8215d0680, 0xc8216cbd60)
	/usr/src/docker/.gopath/src/github.com/docker/docker/container/monitor.go:158 +0xaa
github.com/docker/docker/container.(*containerMonitor).start(0xc8215d0680, 0x0, 0x0)
	/usr/src/docker/.gopath/src/github.com/docker/docker/container/monitor.go:239 +0xbc5
github.com/docker/docker/container.(*containerMonitor).(github.com/docker/docker/container.start)-fm(0x0, 0x0)
	/usr/src/docker/.gopath/src/github.com/docker/docker/container/monitor.go:101 +0x2e
github.com/docker/docker/pkg/promise.Go.func1(0xc8216b2ae0, 0xc821652280)
	/usr/src/docker/.gopath/src/github.com/docker/docker/pkg/promise/promise.go:8 +0x1d
created by github.com/docker/docker/pkg/promise.Go
	/usr/src/docker/.gopath/src/github.com/docker/docker/pkg/promise/promise.go:9 +0x5d

--- a/etchosts/etchosts.go
+++ b/etchosts/etchosts.go
@@ -159,6 +159,10 @@ func Delete(path string, recs []Record) error {
 loop:
 	for s.Scan() {
 		b := s.Bytes()
+		if len(b) == 0 {
+			continue
+		}
+
 		if b[0] == '#' {
 			buf.Write(b)
 			buf.Write(eol)

--- a/etchosts/etchosts_test.go
+++ b/etchosts/etchosts_test.go
@@ -331,6 +331,29 @@ func TestDeleteEmpty(t *testing.T) {
 	}
 }
 
+func TestDeleteNewline(t *testing.T) {
+	file, err := ioutil.TempFile("", "")
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer os.Remove(file.Name())
+
+	b := []byte("\n")
+	if _, err := file.Write(b); err != nil {
+		t.Fatal(err)
+	}
+
+	rec := []Record{
+		{
+			Hosts: "prefix",
+			IP:    "2.2.2.2",
+		},
+	}
+	if err := Delete(file.Name(), rec); err != nil {
+		t.Fatal(err)
+	}
+}
+
 func TestDelete(t *testing.T) {
 	file, err := ioutil.TempFile("", "")
 	if err != nil {

